sql/src/main/java/io/crate/operation/merge/PagingIteratorEmitter.java

```diff
@@ -0,0 +1,145 @@
+     *                         The argument is the "exhaustedBucket" (or null if unknown/all are exhausted).
+     *
+     *                         The function must return true if the fetchMore request works, false if it's not possible
+     *                         to fetch more data.
+     *
+     * @param closeCallback     Callback that is called if all data has been consumed or the rowReceiver returned STOP.
+     *                          If an error occurred during setNextRow this callback is called with a throwable.
+     */
+    public PagingIteratorEmitter(PagingIterator<TKey, Row> pagingIterator,
+                                 RowReceiver rowReceiver,
+                                 Function<TKey, Boolean> fetchMoreTrigger,
+                                 Consumer<Throwable> closeCallback,
+                                 Executor executor) {
+        this.pagingIterator = pagingIterator;
+        this.rowReceiver = rowReceiver;
+        this.fetchMoreTrigger = fetchMoreTrigger;
+        this.closeCallback = closeCallback;
+        this.executor = executor;
+    }
+
+    /**
+     * Consumes the {@link PagingIterator} and pushes the rows into {@link RowReceiver}.
+     *
+     * Whoever uses this emitter must make sure to provide the PagingIterator with data using {@link PagingIterator#merge(Iterable)}
+     * before calling this function.
+     */
+    public void consumeItAndFetchMore() {
+        if (consumeIt() == RowReceiver.Result.CONTINUE) {
+            Boolean fetchMoreSucceeded = fetchMoreTrigger.apply(pagingIterator.exhaustedIterable());
+            upstreamExhausted = !fetchMoreSucceeded;
+            if (upstreamExhausted) {
+                consumeRemaining();
+            }
+        }
+    }
+
+    private void consumeRemaining() {
+        pagingIterator.finish();
+        if (consumeIt() == RowReceiver.Result.CONTINUE) {
+            // can't continue, no more rows
+            rowReceiver.finish(this::repeat);
+            closeCallback.accept(null);
+        }
+    }
+
+    private void repeat() {
+        IterableRowEmitter emitter = new IterableRowEmitter(rowReceiver, pagingIterator.repeat(), Optional.of(executor));
+        emitter.run();
+    }
+
+    private RowReceiver.Result consumeIt() {
+        while (pagingIterator.hasNext()) {
+            Row row = pagingIterator.next();
+            RowReceiver.Result result;
+            try {
+                result = rowReceiver.setNextRow(row);
+            } catch (Throwable t) {
+                return propagateFailAndStop(t);
+            }
+            switch (result) {
+                case CONTINUE:
+                    continue;
+
+                case PAUSE:
+                    rowReceiver.pauseProcessed(this::resume);
+                    return result;
+
+                case STOP:
+                    return propagateFinishAndStop(result);
+            }
+        }
```

Comment:
in general I agree with @mfussenegger that using an enum for states is much more obvious than a simple boolean. I also understand @dobe's point of abusing the Result enum here for internal purpose but defining an internal enum for the sake of keeping it private would be a bit over-engineering imho. I think it is ok and clear even if only the CONTINUE state is explicit handled, it is clear to me that nothing has to be done for the other states. just a boolean wouldn't make is such obvious imho. (we even replaced some booleans in the past with a 2-value enum just to make the intent more obvious)